### PR TITLE
[NEW] Add functionality to allow user to create new chat rooms

### PR DESCRIPTION
**Pull request by aniketsingh03**
@RocketChat/android

Closes https://github.com/RocketChat/Rocket.Chat.Android/issues/857

This pr aims to complete the following features:
- [x] Add FAB to create new channel.
- [x] Add a view to create a new channel.
- [x] Allow creation of new public channels and private groups
- [x] Call the rest api function to successfully create a new channel 
- [x] Add a view to invite members to the new channel.
- [x] Send list of members along with channel name in the rest call.
- [x] Support infinite scrolling for search results.
- [x] Fix CI tests

### CODE REVIEW

**Comment by luciofm** [comment]
Don't think a FAB is the right choice for creating new rooms, it is not a very common action that deserves a FAB for it...

**Comment by aniketsingh03** [comment]
I agree with you @luciofm, but FAB is usually used for positive actions like creating, sharing etc. Also, I have seen many applications using a FAB for such actions, so I thought that FAB is a good choice. Can you suggest an alternative for it?

**Comment by mvp-sales** [comment]
@aniketsingh03 Are you still working in this issue?

**Comment by aniketsingh03** [comment]
Yeah, I'm close to completing it, I had been busy for some days so wasn't able to work on it, but I'll complete it soon.

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/presentation/AddMembersPresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/presentation/AddMembersPresenter.kt]: Just declare  `private val client = factory.create(serverInteractor.get()!!)` and remove `val serverUrl: String? = serverInteractor.get()`.

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt:166]: 💯 👍

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Variable never changes, so use `val` instead of `var` here.

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelPresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelPresenter.kt]: `channelCreated` ins't being used, so remove this attribution.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: `ShowMessageAndClearTest()`? It is better to create another function and avoid doing things that a function isn't made for.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: `ShowMessageAndClearTest()`? It is better to create another function and avoid doing things that a function isn't made for.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: `ShowMessageAndClearTest()`? It is better to create another function and avoid doing things that a function isn't made for.

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Create a function for this specific thing too.

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Need to dispose the subscription too.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Need to dispose the subscription too.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Looking at this class I can see that there are `"public", "c" and "p"` ... wait! `public` does not comply with our API! Also, use `lateinit` here and why not specifying the room type using the proper model instead of String?

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Why not specifying the room type using the proper model instead of String?

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Why not specifying the room type using the proper model instead of String?

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: You could define `resources.getColor(R.color.colorRed)` in the same line here.

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (aniketsingh03) [reply app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Yeah, thanks for pointing this out.

**Comment by aniketsingh03** [comment]
@filipedelimabrito please have a look.

**Comment by aniketsingh03** [comment]
@filipedelimabrito please have a look if this can be merged.

File: app/build.gradle
Comment (philipbrito) [inline app/build.gradle]: Re add the blank line here please. If you take a look, we a adding blank space to separe the libraries by its `artifactId`

File: app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomFragment.kt]: We don't need `with` here. You can change to `text_message.text ...`. If you could do this it will be better.

File: app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt]: Everything is `setup[something]` so why `setUp[FAB]`. Also, change to `Fab` instead of `FAB`

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: You can use a class function here: `showMessage()`

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Better to dispose on `onStop()` lifecycle. But attempt to `setUpObservableForSearchView()` on `onStart()`.

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Change to `progress_bar.isVisible = true`

File: app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/addMembers/ui/AddMembersActivity.kt]: Change to `progress_bar.isVisible = false`

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelPresenter.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelPresenter.kt:13]: Remove this unnecessary blank line.

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt]: For adding the doc you should use 
```
/**
 *
/**
```
 
Take a look on some app interface functions for reference.

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt]: For adding the doc you should use 
```
/**
 *
/**
```
 
Take a look on some app interface functions for reference.

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt]: For adding the doc you should use 
```
/**
 *
/**
```
 
Take a look on some app interface functions for reference.

File: app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/presentation/CreateNewChannelView.kt]: Add a doc here too if you can.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: The same case that I wrote above about disposing the observable in another lifecycle function. Please, take a look.

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Change `view_loading.setVisible(true)` to `view_loading.isVisible =  true`

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt]: Change `view_loading.setVisible(false)` to `view_loading.isVisible =  false`

File: app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/createChannel/ui/CreateNewChannelActivity.kt:90]: Remove this blank line here please.

File: app/src/main/res/layout/activity_add_members.xml
Comment (philipbrito) [inline app/src/main/res/layout/activity_add_members.xml]: Better to define it as `view_loading` as you, correctly, define on activity_create_new_channel.xml.

**Review by philipbrito — CHANGES_REQUESTED** [review]

**Comment by aniketsingh03** [comment]
@filipedelimabrito please have a look

**Review by philipbrito — CHANGES_REQUESTED** [review]
Feature is incomplete for me. I could not get results for the members I was searching for.

Another improvement that needs to be addressed before the merge is the color of the FAB. You can use the `colorAccent` - that is what the Google design guidelines recommends. Btw, this could be an item in the nav. drawer (that suits better IMHO).

**Comment by philipbrito** [comment]
Here an example what I'm saying above:

**Comment by philipbrito** [comment]
Also, I was unable to create a chatroom because I have not added team members. Why?

**Comment by aniketsingh03** [comment]
@filipedelimabrito This is happening because at https://github.com/RocketChat/Rocket.Chat.Android/blob/45f52fb70d0b010d5147ee053aed09715f867c85/app/src/main/java/chat/rocket/android/members/viewmodel/MemberViewModel.kt#L35, `settings.useRealName()` returns `false` so the username is set which is eventually null for the user you have searched!! I guess this function hasn't been implemented correctly, it should rather return the value which is not null after checking for settings, otherwise we can get null in many cases and that will be displayed as a string.

**Comment by philipbrito** [comment]
So, fix that please. We can't ship it in that way. Also, you bring another question: Is it searching by username or real name?

Also, be aware that we are using this function and it is working properly in other cases. Better, we are able to see the real name and username in the member view correctly.

**Comment by aniketsingh03** [comment]
@filipedelimabrito Apparently, both the api's (member details and user query) return different results and on searching for filipe, the user query api returns only one user ie the one you searched for whereas the user returned from member details is a different user which is never returned in the case of querying users. It's strange but this is what is happening! However, I'll modify the function to not return a null value.

File: app/src/main/java/chat/rocket/android/members/viewmodel/MemberViewModel.kt
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/members/viewmodel/MemberViewModel.kt:39]: It will not respect the settings. This is wrong. Please, rollback it.

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/java/chat/rocket/android/members/viewmodel/MemberViewModel.kt
Comment (aniketsingh03) [reply app/src/main/java/chat/rocket/android/members/viewmodel/MemberViewModel.kt:39]: What should we do in case the settings do not allow real name and the username is null?

**Pull request merged**
